File: ios/ClaudeCursor/ClaudeCursor/Services/APIClient.swift
import Foundation

enum APIError: LocalizedError {
    case invalidURL
    case httpError(Int, url: String)
    case decodingError(String)

    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Invalid URL"
        case .httpError(let code, let url):
            return "HTTP \(code): \(url)"
        case .decodingError(let detail):
            return "Decode error: \(detail)"
        }
    }
}

actor APIClient {
    let baseURL: String
    private let session: URLSession

    init(baseURL: URL) {
        self.baseURL = baseURL.absoluteString.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        let config = URLSessionConfiguration.default
        // Use a browser-like User-Agent so Cloudflare Tunnel doesn't reject requests
        config.httpAdditionalHeaders = [
            "User-Agent": "Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.0 Mobile/15E148 Safari/604.1 ClaudeCursorCompanion/1.0"
        ]
        self.session = URLSession(configuration: config)
    }

    private func url(_ path: String) -> URL? {
        URL(string: "\(baseURL)\(path)")
    }

    private func request<T: Decodable>(_ method: String, path: String, body: Data? = nil) async throws -> T {
        guard let url = url(path) else { throw APIError.invalidURL }
        var req = URLRequest(url: url)
        req.httpMethod = method
        if let body {
            req.httpBody = body
            req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }
        let (data, response) = try await session.data(for: req)
        guard let http = response as? HTTPURLResponse else { throw APIError.invalidURL }
        guard (200...299).contains(http.statusCode) else {
            let preview = String(data: data, encoding: .utf8)?.prefix(200) ?? ""
            throw APIError.httpError(http.statusCode, url: "\(url.absoluteString) | \(preview)")
        }
        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        do {
            return try decoder.decode(T.self, from: data)
        } catch {
            let preview = String(data: data, encoding: .utf8)?.prefix(300) ?? ""
            throw APIError.decodingError("\(error.localizedDescription) | \(preview)")
        }
    }

    private func requestVoid(_ method: String, path: String, body: Data? = nil) async throws {
        guard let url = url(path) else { throw APIError.invalidURL }
        var req = URLRequest(url: url)
        req.httpMethod = method
        if let body {
            req.httpBody = body
            req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }
        let (data, response) = try await session.data(for: req)
        guard let http = response as? HTTPURLResponse else { throw APIError.invalidURL }
        guard (200...299).contains(http.statusCode) else {
            let preview = String(data: data, encoding: .utf8)?.prefix(200) ?? ""
            throw APIError.httpError(http.statusCode, url: "\(url.absoluteString) | \(preview)")
        }
    }

    func fetchSessions() async throws -> [Session] {
        let response: SessionsResponse = try await request("GET", path: "/api/sessions")
        return response.sessions
    }

    func fetchShells() async throws -> [Shell] {
        let response: ShellsResponse = try await request("GET", path: "/api/sessions/shells")
        return response.shells
    }

    func createSession(name: String?, shell: String) async throws {
        var params: [String: String] = ["shell": shell]
        if let name, !name.isEmpty { params["name"] = name }
        let body = try JSONEncoder().encode(params)
        try await requestVoid("POST", path: "/api/sessions", body: body)
    }

    func stopSession(name: String) async throws {
        try await requestVoid("POST", path: "/api/sessions/\(name)/stop")
    }

    func restartSession(name: String) async throws {
        try await requestVoid("POST", path: "/api/sessions/\(name)/restart")
    }

    func deleteSession(name: String) async throws {
        try await requestVoid("DELETE", path: "/api/sessions/\(name)")
    }

    func resolveShareToken(_ token: String) async throws -> ShareInfo {
        return try await request("GET", path: "/api/sessions/share/\(token)")
    }
}

File: android/ClaudeCursor/app/src/main/java/com/claudecursor/app/data/remote/ApiClient.kt
package com.claudecursor.app.data.remote

import com.claudecursor.app.data.remote.dto.*
import com.claudecursor.app.domain.model.Session
import com.claudecursor.app.domain.model.Shell
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.http.*
import java.util.concurrent.TimeUnit

private const val USER_AGENT =
    "Mozilla/5.0 (Linux; Android 14; Pixel 8) AppleWebKit/537.36 (KHTML, like Gecko) " +
    "Chrome/120.0.0.0 Mobile Safari/537.36 ClaudeCursorCompanion/1.0"

interface ApiService {
    @GET("/api/sessions")
    suspend fun getSessions(): SessionsResponse

    @GET("/api/sessions/shells")
    suspend fun getShells(): ShellsResponse

    @POST("/api/sessions")
    suspend fun createSession(@Body body: CreateSessionRequest)

    @POST("/api/sessions/{name}/stop")
    suspend fun stopSession(@Path("name") name: String)

    @POST("/api/sessions/{name}/restart")
    suspend fun restartSession(@Path("name") name: String)

    @DELETE("/api/sessions/{name}")
    suspend fun deleteSession(@Path("name") name: String)

    @GET("/api/sessions/share/{token}")
    suspend fun resolveShareToken(@Path("token") token: String): ShareInfoDto
}

object ApiClient {
    private val clients = mutableMapOf<String, ApiService>()

    fun getService(baseURL: String): ApiService {
        return clients.getOrPut(baseURL) {
            val client = OkHttpClient.Builder()
                .addInterceptor(Interceptor { chain ->
                    val request = chain.request().newBuilder()
                        .header("User-Agent", USER_AGENT)
                        .build()
                    chain.proceed(request)
                })
                .connectTimeout(10, TimeUnit.SECONDS)
                .readTimeout(30, TimeUnit.SECONDS)
                .build()

            val url = if (baseURL.endsWith("/")) baseURL else "$baseURL/"

            Retrofit.Builder()
                .baseUrl(url)
                .client(client)
                .addConverterFactory(GsonConverterFactory.create())
                .build()
                .create(ApiService::class.java)
        }
    }

    suspend fun fetchSessions(baseURL: String): List<Session> {
        return getService(baseURL).getSessions().sessions.map { it.toDomain() }
    }

    suspend fun fetchShells(baseURL: String): List<Shell> {
        return getService(baseURL).getShells().shells.map { it.toDomain() }
    }

    suspend fun createSession(baseURL: String, name: String?, shell: String) {
        getService(baseURL).createSession(CreateSessionRequest(name = name, shell = shell))
    }

    suspend fun stopSession(baseURL: String, name: String) {
        getService(baseURL).stopSession(name)
    }

    suspend fun restartSession(baseURL: String, name: String) {
        getService(baseURL).restartSession(name)
    }

    suspend fun deleteSession(baseURL: String, name: String) {
        getService(baseURL).deleteSession(name)
    }

    suspend fun resolveShareToken(baseURL: String, token: String): ShareInfoDto {
        return getService(baseURL).resolveShareToken(token)
    }
}
